Fix TS minimap terrain colours

pchote opened the pull request:
My original interpretation was quite wrong... I don't know what I was thinking 😅 
From the commit message:
```
* Rename LeftColor and RightColor to MinColor and MaxColor
  These are mapped from LowRadarColor and HighRadarColor in
  the original inis, and appear to be used to set the bounding
  values for selecting a random colour, NOT for left/right
  pixels (which caused noticeably wrong banding).

* Adjust brightness based on terrain height.
  MinHeightColorBrightness and MaxHeightColorBrightness
  were chosen by trial/error to match the original
  map preview rendering.
```
Thanks to tomsons26 for explaining that the original games hardcode tileset-specific brightness bodges in the exe. I don't know what those bodges actually *are*, so I picked modifiers that matched as closely as I could tell by eye.

An obvious scope creep will be to replace `TerrainTypeInfo.Color` with `MinColor`/`MaxColor` and define appropriate ranges for RA/TD/D2k's terrain types to add visual noise and break up the large constant color blocks there too. That will mean regenerating all of the upstream oramaps, so I would rather not do that here where the focus is on TS.

Some before/original TS/after comparisons:

![arivruns-old](https://user-images.githubusercontent.com/167819/84711163-91bb1e00-af5d-11ea-85be-e7a485cc75dc.png)
<img width="155" alt="arivruns-minimap" src="https://user-images.githubusercontent.com/167819/84711211-b57e6400-af5d-11ea-9743-b5b30de21083.png">
![arivruns-new](https://user-images.githubusercontent.com/167819/84711220-bb744500-af5d-11ea-951f-129f49dadfc2.png)

![float-old](https://user-images.githubusercontent.com/167819/84712436-7ac9fb00-af60-11ea-98a7-7f6b10f02126.png)
<img width="168" alt="float-minimap" src="https://user-images.githubusercontent.com/167819/84712444-7d2c5500-af60-11ea-89c0-cf270460cf5f.png">
![float-new](https://user-images.githubusercontent.com/167819/84712447-7f8eaf00-af60-11ea-8b0e-e44494858090.png)

![tread_l-old](https://user-images.githubusercontent.com/167819/84711517-7d2b5580-af5e-11ea-8e8a-5fc82b17c5f7.png)
<img width="162" alt="tread_l-minimap" src="https://user-images.githubusercontent.com/167819/84712062-a0a2d000-af5f-11ea-8ed3-b3dd1268e7ec.png">
![tread_l-new](https://user-images.githubusercontent.com/167819/84711528-8288a000-af5e-11ea-99ac-011118b4b913.png)

Mailaender commented on OpenRA.Mods.Common/UpdateRules/Rules/20200503/UpdateTilesetColors.cs:
```diff
@@ -0,0 +1,50 @@
+		{
+			get
+			{
+				return "The LeftColor and RightColor keys in isometric tilesets have been renamed to MinColor and MaxColor to reflect their proper usage.";
```
Although maybe not intended you can also use them in rectangular maps. https://github.com/OpenHV/OpenHV/commit/da97b59f4eeaf4701cfe8fc492a7e14edd46075a

![image](https://user-images.githubusercontent.com/756669/85066507-6465ae80-b1af-11ea-82d2-fd64893109ac.png)

where they give an illusion of depth without me having to define yet another terrain type.

Mailaender commented on OpenRA.Game/Map/Map.cs:
```diff
@@ -716,8 +722,21 @@ public byte[] SavePreview()
+									leftColor = Color.FromArgb((int)(scale * left.R).Clamp(0, 255), (int)(scale * left.G).Clamp(0, 255), (int)(scale * left.B).Clamp(0, 255));
+									rightColor = Color.FromArgb((int)(scale * right.R).Clamp(0, 255), (int)(scale * right.G).Clamp(0, 255), (int)(scale * right.B).Clamp(0, 255));
+								}
+								else
+									leftColor = rightColor = type.MinColor;
+							}
+							else
+								leftColor = rightColor = Color.Black;
```
Shouldn't this be

```suggestion
								else
								{
									leftColor = type.MinColor;
									rightColor = type.MaxColor;
								}
							}
							else
								leftColor = rightColor = Color.Black;
```

?

pchote replied:
No. The two color values do not correspond to the left and right pixels in the isometric minimaps - this was bogus in the previous implementation.

Mailaender commented:
You forgot about https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Mods.Cnc/UtilityCommands/LegacyTilesetImporter.cs but otherwise the result speaks for itself. It looks way better.

pchote committed:
Rewrite TS minimap rendering:

* Rename LeftColor and RightColor to MinColor and MaxColor
  These are mapped from LowRadarColor and HighRadarColor in
  the original inis, and appear to be used to set the bounding
  values for selecting a random colour, NOT for left/right
  pixels (which caused noticeably wrong banding).

* Adjust brightness based on terrain height.
  MinHeightColorBrightness and MaxHeightColorBrightness
  were chosen by trial/error to match the original
  map preview rendering.

pchote commented:
Fixed.

teinarss commented on OpenRA.Game/Map/Map.cs:
```diff
@@ -17,6 +17,7 @@
 using System.Text;
 using OpenRA.FileFormats;
 using OpenRA.FileSystem;
+using OpenRA.Graphics;
```
not needed.

teinarss commented on OpenRA.Game/Map/Map.cs:
```diff
@@ -716,8 +722,21 @@ public byte[] SavePreview()
-							rightColor = type != null ? type.RightColor : Color.Black;
+							if (type != null)
+							{
+								if (useHeightModifiers)
+								{
+									var left = Exts.ColorLerp(Game.CosmeticRandom.NextFloat(), type.MinColor, type.MaxColor);
+									var right = Exts.ColorLerp(Game.CosmeticRandom.NextFloat(), type.MinColor, type.MaxColor);
+									var scale = float2.Lerp(minHeightBrightness, maxHeightBrightness, heightModifier * Height[uv]);
+									leftColor = Color.FromArgb((int)(scale * left.R).Clamp(0, 255), (int)(scale * left.G).Clamp(0, 255), (int)(scale * left.B).Clamp(0, 255));
+									rightColor = Color.FromArgb((int)(scale * right.R).Clamp(0, 255), (int)(scale * right.G).Clamp(0, 255), (int)(scale * right.B).Clamp(0, 255));
+								}
+								else
+									leftColor = rightColor = type.MinColor;
```
Extract this code and reuse it in `PlayerRadarTerrain.cs ` to remove duplication?

pchote committed:
Reduce duplication between Map and PlayerRadarTerrain.

pchote commented:
Updated.

obrakmann reviewed (approved):
:+1:
